Tools/WebKitTestRunner/TestController.cpp

```diff
@@ -360,8 +361,27 @@ void TestController::setIsMediaKeySystemPermissionGranted(bool granted)
     m_isMediaKeySystemPermissionGranted = granted;
 }
 
-static void queryPermission(WKStringRef, WKSecurityOriginRef, WKQueryPermissionResultCallbackRef callback)
+static void queryPermission(WKStringRef string, WKSecurityOriginRef securityOrigin, WKQueryPermissionResultCallbackRef callback)
 {
+    TestController::singleton().handleQueryPermission(string, securityOrigin, callback);
+}
+
+void TestController::handleQueryPermission(WKStringRef string, WKSecurityOriginRef securityOrigin, WKQueryPermissionResultCallbackRef callback)
+{
+    if (toWTFString(string) == "notifications"_s) {
+        auto securityOriginString = adoptWK(WKSecurityOriginCopyToString(securityOrigin));
+        auto dictionary = adoptWK(m_webNotificationProvider.notificationPermissions());
```

Comment:
This is a pretty bad bug. You shouldn't be adopting the returned dictionary since it didn't come from a copy/create function. You need to ref the dictionary if you want to extend its lifetime.

Reply:
We looked at the code and decided not to change this since WebNotificationProvide::notificationPermissions() calls retain on m_permissions.get(), so the adopt is needed. But I've changed the implementation to not call this WebNotificationProvide::notificationPermissions().